File: day15/day15.py
def run(data, count):
    mem   = {}
    idx   = 0
    prev  = 0

    for i in data:
        mem[i] = [idx, idx]
        idx   += 1
        prev   = i

    while idx < count:
        if mem[prev][0] == mem[prev][1]:
            prev = 0
        else:
            last = mem[prev][1] - mem[prev][0]
            prev = last

        if prev in mem:
            mem[prev][0] = mem[prev][1]
            mem[prev][1] = idx
        else:
            mem[prev] = [idx, idx]    
        idx += 1

    return prev 

data = [19,20,14,0,9,1]
print("Part 1:", run(data, 2020))
print("Part 2:", run(data, 30000000))

File: day15/day15.cpp
#include <iostream>
#include <vector>
#include <chrono>


using namespace std;


int run(const vector<uint32_t>& v, uint32_t count)
{
    auto start = chrono::system_clock::now();

    // Don't put the last item into mem. -1 means the 
    // index hasn't been seen yet.
    vector<uint32_t> mem(count, -1);
    for (uint32_t i = 0; i < v.size() - 1; ++i)
    {
        mem[v[i]] = i;
    }

    // The last item is not yet in mem, so we can check if 
    // it is its first appearance.
    uint32_t idx = static_cast<uint32_t>(v.size());
    uint32_t val = v[idx - 1];

    while (idx < count)
    {
        uint32_t next = (mem[val] == -1) ? 0 : idx - mem[val] - 1;
        mem[val] = idx - 1;
        val = next;
        ++idx;
    }

    auto end = chrono::system_clock::now();
    chrono::duration<double> diff = end - start;
    cout << (diff.count() * 1000) << "ms\n";

    return val;    
}


int main()
{
    cout << "Part1: " << run({19, 20, 14, 0, 9, 1}, 2020) << '\n';
    cout << "Part2: " << run({19, 20, 14, 0, 9, 1}, 30'000'000) << '\n';
}